core/src/odbc_uri.rs

```diff
@@ -148,36 +152,70 @@ impl<'a> ODBCUri<'a> {
-                ODBCError::InvalidUriFormat(format!(
+                Error::InvalidUriFormat(format!(
                     "{} is required for a valid Mongo ODBC Uri",
                     names[0]
                 ))
             } else {
-                ODBCError::InvalidUriFormat(format!(
+                Error::InvalidUriFormat(format!(
                     "One of {:?} is required for a valid Mongo ODBC Uri",
                     names
                 ))
             }
         })
     }
 
-    // remove_to_mongo_uri converts this ODBCUri to a mongo_uri String. It will
+    // try_into_client_options converts this ODBCUri to a mongo_uri String. It will
     // remove all the attributes necessary to make a mongo_uri. This is destructive!
-    pub fn remove_to_mongo_uri(&mut self) -> Result<String> {
+    pub fn try_into_client_options(&mut self) -> Result<ClientOptions> {
+        let uri = self.remove(URI);
+        if let Some(uri) = uri {
+            return self.handle_uri(uri);
+        }
+        self.handle_no_uri()
+    }
+
+    fn handle_uri(&mut self, uri: &str) -> Result<ClientOptions> {
+        let mut client_options = ClientOptions::parse(uri)?;
+        if client_options.credential.is_some() {
+            let user = self.remove(USER);
+            if user.is_some() {
+                client_options.credential.as_mut().unwrap().username = user.map(String::from);
+            }
+            let pwd = self.remove(PWD);
+            if pwd.is_some() {
+                client_options.credential.as_mut().unwrap().password = pwd.map(String::from);
+            }
+            return Ok(client_options);
+        }
+        let user = self.remove(USER);
+        let pwd = self.remove(PWD);
+        client_options.credential = Some(
+            Credential::builder()
+                .username(user.map(String::from))
+                .password(pwd.map(String::from))
+                .build(),
+        );
+        Ok(client_options)
+    }
+
+    fn handle_no_uri(&mut self) -> Result<ClientOptions> {
         let user = self.remove_mandatory_attribute(USER)?;
         let pwd = self.remove_mandatory_attribute(PWD)?;
         let server = self.remove_mandatory_attribute(SERVER)?;
         let ssl = self.remove(SSL);
-        let ssl_string =
-            if ssl.is_some() && ssl.unwrap() != "0" && ssl.unwrap().to_lowercase() != "false" {
-                "?ssl=true"
-            } else {
-                ""
-            };
-        Ok(format!(
-            "mongodb://{}:{}@{}{}",
-            user, pwd, server, ssl_string
-        ))
+        let cred = Credential::builder()
+            .username(user.to_string())
+            .password(pwd.to_string())
+            .build();
+        let mut client_options = ClientOptions::builder()
+            .hosts(vec![ServerAddress::parse(server)?])
+            .credential(cred)
+            .build();
+        if ssl.is_some() && ssl.unwrap() != "0" && ssl.unwrap().to_lowercase() != "false" {
```

Comment:
In the connector we won't be going down this path since it will be setting only the URI, driver, and credentials. 
Since we are allowing users to use the other connection string options if the URI isn't set, it does makes sense to include the ODBC connection settings that we know users will want to set: ssl and connectTimeoutMS.  For users who want more customizing we can say to use the URI option.

Reply:
That works for me! I personally don't know which ones users will want except ssl and connectTimeoutMS so I am fine to start with this.
Thank you @bucaojit for confirming :)

Reply:
The SSL one is a hold over from the BIC odbc driver. I am fine removing the option if we want :)

Should probably consider supporting all the options we did before?

Reply:
Ok, removing these now!